scipy/optimize/_lsap_module.c

```diff
@@ -44,9 +44,13 @@ calculate_assignment(PyObject* self, PyObject* args)
         return NULL;
 
     PyArrayObject* obj_cont =
-      (PyArrayObject*)PyArray_ContiguousFromAny(obj_cost, NPY_DOUBLE, 2, 2);
+      (PyArrayObject*)PyArray_ContiguousFromAny(obj_cost, NPY_DOUBLE, 0, 0);
     if (!obj_cont) {
-        PyErr_SetString(PyExc_TypeError, "invalid cost matrix object");
+        return NULL;
+    }
+
+    if (PyArray_NDIM(obj_cont) != 2) {
+        PyErr_SetString(PyExc_TypeError, "cost matrix must be 2D");
```

Comment:
IMO Value error makes sense because the type is `ndarray` and it's attribute `ndim` holds an invalid value. Either way, you can improve the error message with `PyErr_Format`.

```suggestion
        return PyErr_Format(PyExc_ValueError, "expected a matrix (2-D array), got a %d array", PyArray_NDIM(obj_cont));
```

Reply:
Thanks for the suggestion @peterbell10. That is much nicer.